### PR TITLE
Improve big editor

### PR DESCRIPTION
**Pull request by ChugunovRoman**
New features:
- [x] - Implement dialog modal for open/save/import/export file/files
- [x] - Export all files from a big archive with creating full path to exported file
- [x] - Add the scroll for list of files by UpArrow and DownArrow keys
- [x] - Display formated file size
- [x] - Add status bar

- [ ] - Implement import a file or directory with files
- [ ] - Add CLI
- [ ] - Add "Save as..." item in the File menu

### CODE REVIEW

File: src/OpenSage.FileFormats.Big/BigArchive.cs [hunk 1]
```diff
@@ -20,6 +20,7 @@ public class BigArchive : DisposableBase
         internal Stream Stream => _stream;
 
         public string FilePath { get; }
+        public long Size { get; }
```
Comment (feliwir) [inline src/OpenSage.FileFormats.Big/BigArchive.cs]: You could probably let that point at the stream directly like so:
```csharp
public long Size => _stream.Length
```

File: src/OpenSage.FileFormats.Big/BigArchiveEntry.cs [hunk 2]
```diff
@@ -10,6 +10,9 @@ public class BigArchiveEntry
         public BigArchive Archive { get; }
 
         public string FullName { get; }
+        public string Name {
+            get { return FullName.Split('\\')[FullName.Split('\\').Length - 1]; }
```
Comment (feliwir) [inline src/OpenSage.FileFormats.Big/BigArchiveEntry.cs]: You could probably don't want to split twice here. Better use an intermediate variable

File: src/OpenSage.Tools.BigEditor/Program.cs [hunk 3]
```diff
@@ -1,6 +1,6 @@
-﻿using OpenSage.Tools.BigEditor.UI;
-using Veldrid;
+﻿using Veldrid;
 using Veldrid.StartupUtilities;
+using OpenSage.Tools.BigEditor.UI;
```
Comment (feliwir) [inline src/OpenSage.Tools.BigEditor/Program.cs]: Why is that required now?

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 4]
```diff
@@ -59,14 +66,18 @@ private void UpdateSearch(string searchText)
 
         public void Draw(Sdl2Window window)
         {
+            _window = window;
```
Comment (feliwir) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Why you store the window as a member? I don't see it being used somewhere

Comment (feliwir) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Okay, i found it but better pass the width & height to `DrawFileContent`, so we don't need that member

File: src/OpenSage.FileFormats.Big/BigArchiveEntry.cs (hunk 2 above)
Comment (ChugunovRoman) [reply src/OpenSage.FileFormats.Big/BigArchiveEntry.cs]: Do you mean that?
```csharp
public string Name {
            get {
                int itemIndex = FullName.Split('\\').Length - 1;
                return FullName.Split('\\')[itemIndex];
            }
        }
```

Comment (feliwir) [reply src/OpenSage.FileFormats.Big/BigArchiveEntry.cs]:
```csharp
public string Name {
            get {
                var splitted= FullName.Split('\\');
                return splitted[splitted.Length-1];
            }
        }
```

File: src/OpenSage.Tools.BigEditor/Program.cs (hunk 3 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/Program.cs]: It's need for create the main form

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs (hunk 4 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: It used for getting width and height of window in the DrawFilesList function at 355 line

Comment (feliwir) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Yes, better pass a `Vector2` to DrawFilesList with the size. Because the window is only valid during the "scope" of the Draw function. So:
```csharp
 private void DrawFilesList(Vector2 windowSize)
{
.
.
  ImGui.SetWindowSize(windowSize - 50, ImGuiCond.Always);
.
.
}
```

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+            }
+            catch (Exception e) {
+                if (ImGui.TreeNodeEx(label)) {
+                    ImGui.Text(e.Message);
+                    ImGui.TreePop();
+                }
+                return false;
+            }
+
+            var currentSelect = Regex.Match($"{_currentPath}/", $"{path}");
```
Comment (feliwir) [inline src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: Crashes for me:
```
System.ArgumentException: "parsing '\$RECYCLE.BIN\' - Illegal \\ at end of pattern."
```

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs (hunk 4 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Is it possible subtract int from Vector2 ?
it;s different types

Comment (feliwir) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: You need to do `- new Vector2(50)`

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs (hunk 5 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: Maybe the problem in regex, because in this case regex contain not shielded "$" char

Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: Or not. I created $RECYCLE.BIN directory and work fine.
Ok, I'll try to test this on Windows

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+            }
+
+            ImGui.EndChild(); // browser
+
+            ImGui.EndChild(); // container
+
+            return false;
+        }
+
+        private static bool DirTree(string currentPath = "/")
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: The return value is never used.

**Comment by ChugunovRoman** [comment]
> I don't think writing a new file picker from scratch is a good idea - it's a lot of work for not a lot of gain. Your current implementation doesn't work on Windows, because it doesn't allow the user to choose a drive. Mac users have different expectations for a file browser, so we would effectively need to customise the file picker for all 3 platforms.

Well, for now I am working on cross-platform file browser. And it works on Linux and windows. On my laptop, and in VM windows. I don't push this changes yet.
I can continue on cross-platform file browser. But is it necessary? 
I think in general, the implement of CLI without GUI will be enough

**Comment by feliwir** [comment]
I think having a GUI definetly makes sense, since windows users probably won't use the CLI too much. We could probably use your branch without the FileDialog and add the native filedialog later.

File: src/OpenSage.Tools.BigEditor/Program.cs [hunk 7]
```diff
@@ -14,7 +14,7 @@ static void Main(string[] args)
             const int initialHeight = 768;
 
             VeldridStartup.CreateWindowAndGraphicsDevice(
-                new WindowCreateInfo(100, 100, initialWidth, initialHeight, WindowState.Normal, "OpenSAGE Big Editor"),
+                new WindowCreateInfo(500, 100, initialWidth, initialHeight, WindowState.Normal, "OpenSAGE Big Editor"),
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/Program.cs]: I wanted the window displayed in screen center. But I don't know how to do it still, and I set this on 500. But it isn't a center of a screen)

**Comment by ChugunovRoman** [comment]
> I think having a GUI definetly makes sense, since windows users probably won't use the CLI too much. We could probably use your branch without the FileDialog and add the native filedialog later.

Ok. I can disabled my file browser and return old version.
And implement CLI.
In future, can I try to implement native library for native filedialog on C++?
Or would you like to do it yourself?
I'm interested in it

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs (hunk 5 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: Fixed

File: src/OpenSage.FileFormats.Big/BigArchiveEntry.cs [hunk 8]
```diff
@@ -10,6 +10,14 @@ public class BigArchiveEntry
         public BigArchive Archive { get; }
 
         public string FullName { get; }
+        public string Name
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.FileFormats.Big/BigArchiveEntry.cs]: BigArchiveEntry.cs(13,52): error CS1061: 'string[]' does not contain a definition for 'Last' and no accessible extension method 'Last' accepting a first argument of type 'string[]' could be found (are you missing a using directive or an assembly reference?) [/media/ruut/ssd/Projects/otherProjects/OpenSAGE/src/OpenSage.FileFormats.Big/OpenSage.FileFormats.Big.csproj]

File: src/OpenSage.Tools.BigEditor/Program.cs [hunk 9]
```diff
@@ -26,7 +26,7 @@ static void Main(string[] args)
             {
                 window.Resized += () =>
                 {
-                    graphicsDevice.ResizeMainWindow((uint) window.Width, (uint) window.Height);
+                    graphicsDevice.ResizeMainWindow((uint)window.Width, (uint)window.Height);
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/Program.cs]: It's auto format in vscode.

File: src/OpenSage.Tools.BigEditor/Program.cs [hunk 10]
```diff
@@ -50,7 +50,7 @@ static void Main(string[] args)
                         commandList.ClearColorTarget(0, RgbaFloat.Clear);
 
                         imGuiRenderer.Update(
-                            (float) gameTimer.CurrentGameTime.DeltaTime.TotalSeconds,
+                            (float)gameTimer.CurrentGameTime.DeltaTime.TotalSeconds,
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/Program.cs]: Auto format

File: src/OpenSage.FileFormats.Big/BigArchiveEntry.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Qibbi) [reply src/OpenSage.FileFormats.Big/BigArchiveEntry.cs]: It's in System.Linq

[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.FileFormats.Big/BigArchiveEntry.cs]: Fixed

File: src/OpenSage.Tools.BigEditor/Program.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (feliwir) [reply src/OpenSage.Tools.BigEditor/Program.cs]: Please undo this change

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 11]
```diff
@@ -253,53 +299,91 @@ private void DrawFilesList()
...
+                        {
+                            entryStream.CopyTo(fileStream);
+                        }
                     }
-
-                    ImGui.EndPopup();
+                }
+                catch (Exception e)
+                {
+                    Console.WriteLine(e.Message);
```
Comment (feliwir) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: We don't use `Console.WriteLine`. Please look into our builtin logger, see some other files:
`private static NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();`

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+namespace OpenSage.Tools.BigEditor.UI
+{
+    internal sealed class FileBrowser : DisposableBase
+    {
+        private static string _currentPath;
+        private static string _currentFile;
+        private static string _currentExportFileName;
+        private static int _currentSelection;
+        private static FileBrowserType _type;
+        private static Hashtable _paths;
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: Not actual

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+using System.IO;
+using System.Numerics;
+using ImGuiNET;
+using OpenSage.Tools.BigEditor.Util;
+
+namespace OpenSage.Tools.BigEditor.UI
+{
+    internal sealed class FileBrowser : DisposableBase
+    {
+        private static string _currentPath;
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: Not actual

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs [hunk 14]
```diff
@@ -0,0 +1,346 @@
...
+        private static int _currentSelection;
+        private static FileBrowserType _type;
+        private static Hashtable _paths;
+
+
+        public FileBrowser()
+        {
+            _paths = new Hashtable();
+
+            _paths.Add("CNC_GENERALS_PATH", ImGuiUtility.NormalizePath(Environment.GetEnvironmentVariable("CNC_GENERALS_PATH")));
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: Not actual

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs (hunk 6 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: Not actual

File: src/OpenSage.Tools.BigEditor/Program.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/Program.cs]: fixed

File: src/OpenSage.Tools.BigEditor/Util/AdapterEntry.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
+using System;
+
+namespace OpenSage.Tools.BigEditor.Util
+{
+    public class AdapterEntry
+    {
+
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/Util/AdapterEntry.cs]: Not actual

File: src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/FileBrowser.cs]: I removed the fileBrowser

File: src/OpenSage.Tools.BigEditor/Util/ImGuiUtility.cs [hunk 16]
```diff
@@ -1,13 +1,24 @@
 ﻿using System.Text;
+using System.IO;
 using ImGuiNET;
 
 namespace OpenSage.Tools.BigEditor.Util
 {
+    public enum FileBrowserType
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/Util/ImGuiUtility.cs]: I removed this enum

File: src/OpenSage.Tools.BigEditor/Util/Adapter.cs [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace OpenSage.Tools.BigEditor.Util
+{
+    // Combinations files and directories together
+    // in one enumerable object and provides a access via common interface
+    public class Adapter
+    {
+        private readonly List<AdapterEntry> _entries;
+
+        public readonly string RootDirectory = "";
```
[the remark this answers is not in the document]
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/Util/Adapter.cs]: The code was removed

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs (hunk 11 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Replaces on NLog.Logger

File: src/OpenSage.Tools.BigEditor/Util/ImGuiUtility.cs [hunk 18]
```diff
@@ -7,7 +7,7 @@ internal static class ImGuiUtility
     {
         public static unsafe bool InputText(string label, byte[] textBuffer, out string result)
         {
-            var temp = ImGui.InputText(label, textBuffer, (uint) textBuffer.Length, ImGuiInputTextFlags.None, data => 0);
+            var temp = ImGui.InputText(label, textBuffer, (uint)textBuffer.Length, ImGuiInputTextFlags.None, data => 0);
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/Util/ImGuiUtility.cs]: Reverse this whitespace change.

```suggestion
            var temp = ImGui.InputText(label, textBuffer, (uint) textBuffer.Length, ImGuiInputTextFlags.None, data => 0);
```

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 19]
```diff
@@ -253,53 +301,91 @@ private void DrawFilesList()
...
+                {
+                    entryStream.CopyTo(fileStream);
+                }
+            }
+        }
 
-            ImGui.EndChild();
+        private void Import(string directory)
+        {
+            logger.Info("Import files from: {0}", directory);
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Could you add a TODO comment here since this is not implemented yet?

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 20]
```diff
@@ -253,53 +301,91 @@ private void DrawFilesList()
...
+        }
 
-            ImGui.EndChild();
+        private void ImportFile(string filePath)
+        {
+            logger.Info("Import file: {0}", filePath);
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Same here.

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 21]
```diff
@@ -253,53 +301,91 @@ private void DrawFilesList()
...
-                    ImGui.SameLine();
+            ImGui.EndChild(); // end statusbar
+        }
 
-                    if (ImGui.Button("Cancel"))
+        private void Export(string path)
+        {
+            foreach (var entry in _files)
+            {
+                string filePath = $"{path}{Path.DirectorySeparatorChar}{Path.GetDirectoryName(entry.FullName.Replace('\\', Path.DirectorySeparatorChar))}";
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: This is probably a good place to use [`Path.Combine`](https://docs.microsoft.com/en-us/dotnet/api/system.io.path.combine?view=netstandard-2.1)

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 22]
```diff
@@ -15,15 +16,21 @@ internal sealed class MainForm : DisposableBase
...
         private byte[] _filePathBuffer;
         private string _searchText;
+        static float _scrollY;
+
+        private static NLog.Logger logger = NLog.LogManager.GetCurrentClassLogger();
 
         public MainForm()
         {
             _files = new List<BigArchiveEntry>();
+            _currentFileName = "";
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Maybe this should be `null` when no file is selected?

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 23]
```diff
@@ -15,15 +16,21 @@ internal sealed class MainForm : DisposableBase
         private BigArchive _bigArchive;
         private readonly List<BigArchiveEntry> _files;
         private int _currentFile;
+        private string _currentFileName;
         private string _currentFileText;
 
         private byte[] _searchTextBuffer;
         private byte[] _filePathBuffer;
         private string _searchText;
+        static float _scrollY;
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: I don't think this needs to be `static`.

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 24]
```diff
@@ -66,7 +73,10 @@ public void Draw(Sdl2Window window)
             bool open = false;
             ImGui.Begin("OpenSAGE Big Editor", ref open, ImGuiWindowFlags.MenuBar | ImGuiWindowFlags.NoTitleBar | ImGuiWindowFlags.NoResize);
 
-            var wasOpenClicked = false;
+            bool wasOpenClicked = false;
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]:
```suggestion
            var wasOpenClicked = false;
```

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 25]
```diff
@@ -66,7 +73,10 @@ public void Draw(Sdl2Window window)
...
+            bool wasExportAllClicked = false;
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]:
```suggestion
            var wasExportAllClicked = false;
```

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 26]
```diff
@@ -66,7 +73,10 @@ public void Draw(Sdl2Window window)
...
+            bool wasImportFilesClicked = false;
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]:
```suggestion
            var wasImportFilesClicked = false;
```

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 27]
```diff
@@ -66,7 +73,10 @@ public void Draw(Sdl2Window window)
...
+            bool wasImportFileClicked = false;
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]:
```suggestion
            var wasImportFileClicked = false;
```

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs [hunk 28]
```diff
@@ -253,53 +301,91 @@ private void DrawFilesList()
...
+        private void DrawStatusPanel()
+        {
+            ImGui.BeginChild("statusbar", new Vector2(0, 30), true, 0);
+
+            ImGui.Text($"{_bigArchive.FilePath} | Version: {_bigArchive.Version} | Size: {ImGuiUtility.GetFormatedSize(_bigArchive.Size)} | Files: {_bigArchive.Entries.Count}");
+
+            ImGui.SameLine();
 
-                    ImGui.SetItemDefaultFocus();
+            if (_currentFileName.CompareTo("") != 0)
```
Comment (paavohuhtala) [inline src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: You can compare strings with `==` and `!=`. If you make the suggested change to use `null` when no file is selected, then use this:

 ```suggestion
             if (_currentFileName != null)
```

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs (hunk 19 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Added comment

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs (hunk 20 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Added comment

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs (hunk 21 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Replaced

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs (hunk 22 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Maybe) Changed

File: src/OpenSage.Tools.BigEditor/UI/MainForm.cs (hunk 23 above)
Comment (ChugunovRoman) [reply src/OpenSage.Tools.BigEditor/UI/MainForm.cs]: Fixed
